File: valid_parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  q = []
  s.each_char do |c|
    if ['(','{','['].include?(c)
      q << c
    elsif [')','}',']'].include?(c) && closing_bracket(c) == q.last
      q.pop 
    end
  end
end

def closing_bracket(c)
  valid = if c == ')'
    '('
  elsif c == '}'
    '{'
  elsif c == ']'
    '['
  else
    false
  end
  
  valid
end

File: valid-parentheses.js
// O(n)
// S(n)
const isValid = (s) => {
  let [q, dict]=[[], { '}':'{', ']':'[', ')':'(' }]
  for (let i=0; i<s.length; i++) {
    (q.length && dict[s[i]] === q[q.length-1]) ? q.pop() : q.push(s[i]);
  }
  return q.length === 0;
};
